google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -40,14 +40,24 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
 // TODO(#12528): Implement functionality for Flush.
 void TracingMessageBatch::Flush() { child_->Flush(); }
 
-// TODO(#12528): Implement functionality for FlushCallback.
-void TracingMessageBatch::FlushCallback() { child_->FlushCallback(); }
+void TracingMessageBatch::FlushCallback() {
+  for (auto& span : batch_sink_spans_) {
+    internal::EndSpan(*span);
+  }
+  batch_sink_spans_.clear();
+  child_->FlushCallback();
+}
```

Comment:
I think `FlushCallback()` and `SaveMessage()` can be called concurrently by multiple threads?  This code is not thread safe if I am right.

Consider adding a mutex and changing the implementation to something like:

```suggestion
void TracingMessageBatch::FlushCallback() {
  decltype(batch_sink_spans_) spans;
   {
      std::lock_guard<std::mutex> lk(mu_);
      spans_.swap(batch_sink_spans_);
    }
  for (auto& s : spans) internal::EndSpan(*span);
  child_->FlushCallback();
}
```

Reply:
Done. I added a lock around the push_back in SaveMessage too.

Reply:
Is there a good way to test this in a unit test?